Brokerages/Bitfinex/BitfinexBrokerage.Util.cs

```diff
@@ -0,0 +1,164 @@
+            { new OrderTypeMap { BitfinexOrderType = "stop", Wallet = WalletType.trading.ToString(), OrderType = OrderType.StopMarket } },
+            //{ new OrderTypeMap { BitfinexOrderType = "trailing stop", Wallet = WalletType.trading.ToString(), OrderType = OrderType.StopLimit } },
+        };
+        #endregion
+
+
+        /// <summary>
+        /// Map exchange status
+        /// </summary>
+        /// <param name="response"></param>
+        /// <returns></returns>
+        public static OrderStatus MapOrderStatus(BitfinexOrderStatusResponse response)
+        {
+            decimal remainingAmount;
+            decimal executedAmount;
+            if (response.IsCancelled)
+            {
+                return OrderStatus.Canceled;
+            }
+            else if (decimal.TryParse(response.RemainingAmount, out remainingAmount) && remainingAmount > 0
+                && decimal.TryParse(response.ExecutedAmount, out executedAmount) && executedAmount > 0)
+            {
+                return OrderStatus.PartiallyFilled;
+            }
+            else if (response.IsLive)
+            {
+                return OrderStatus.Submitted;
+            }
+
+            return OrderStatus.Invalid;
+        }
+
+
+        /// <summary>
+        /// Map exchange order type
+        /// </summary>
+        /// <param name="orderType"></param>
+        /// <returns></returns>
+        public string MapOrderType(OrderType orderType)
+        {
+            var result = _orderTypeMap.Where(o => o.Wallet == Wallet && o.OrderType == orderType);
+
+            if (result != null & result.Count() == 1)
+            {
+                return result.Single().BitfinexOrderType;
+            }
+
+            throw new Exception("Order type not supported: " + orderType.ToString());
+        }
+
+
+        /// <summary>
+        /// Map exchange order type
+        /// </summary>
+        /// <param name="orderType"></param>
+        /// <returns></returns>
+        public OrderType MapOrderType(string orderType)
+        {
+            var result = _orderTypeMap.Where(o => o.Wallet == Wallet && o.BitfinexOrderType == orderType);
+
+            if (result != null & result.Count() == 1)
+            {
+                return result.Single().OrderType;
+            }
+
+            throw new Exception("Order type not supported: " + orderType.ToString());
+        }
+
+        /// <summary>
+        /// Map exchange order status
+        /// </summary>
+        /// <param name="msg"></param>
+        /// <returns></returns>
+        public OrderStatus MapOrderStatus(TradeMessage msg)
+        {
+            if (Math.Abs(msg.FEE) != 0)
+            {
+                var cached = this.CachedOrderIDs.Where(c => c.Value.BrokerId.Contains(msg.TRD_ORD_ID.ToString())).FirstOrDefault();
+                if (cached.Value != null)
+                {
+                    if (msg.TRD_ORD_ID == cached.Value.BrokerId.Select(b => int.Parse(b)).Max(b => b))
```

Comment:
Calls to `Enumerable.Max(...)` can throw exceptions if there is nothing in the enumerable, so in this case, if `cached.Value.BrokerId` is empty it could throw, also, if you're using `Max(b => b)` you should be able to just write it as `Max()`

In this case, I think `cached.Value.BrokerId` is guaranteed to be populated before it makes its way into the cache. If it is not guaranteed, you can get around the exception by using the `DefaultIfEmpty` method like the following:
```c#
enumerable.DefaultIfEmpty(42).Max();
```
This will find the max value in the enumerable, but in the event that the enumerable is empty then it will return the default value of 42.